File: experimental/csharp/GsAvatar.Runtime/Stream.cs
namespace GsAvatar;

public sealed class GsAvatarSnapshot
{
    public byte Protocol { get; set; } = Constants.ProtocolU8;
    public byte StreamLod { get; set; } = Constants.LodMedium;
    public uint Seq { get; set; }
    public uint CaptureTick { get; set; }
    public ushort Flags { get; set; } = Constants.HasRoot | Constants.HasFace | Constants.HasHands | Constants.HasExpr;
    public double[] RootPos { get; set; } = new double[3];
    public double[] RootRot { get; set; } = { 0, 0, 0, 1 };
    public double[] Body { get; set; } = new double[Constants.BodyDim];
    public double[] LeftHand { get; set; } = new double[Constants.HandDim];
    public double[] RightHand { get; set; } = new double[Constants.HandDim];
    public double[] Expr { get; set; } = new double[Constants.ExprFull];

    public SmplxPose ToPose()
    {
        var expr = new double[Constants.ExprFull];
        Array.Copy(Expr, expr, Math.Min(Expr.Length, expr.Length));
        return new SmplxPose
        {
            Body = (double[])Body.Clone(),
            LeftHand = (double[])LeftHand.Clone(),
            RightHand = (double[])RightHand.Clone(),
            Expr = expr,
            RootPos = (double[])RootPos.Clone(),
            RootRot = (double[])RootRot.Clone(),
            CaptureTick = CaptureTick,
            Teleport = (Flags & Constants.Teleport) != 0,
        };
    }

    public static GsAvatarSnapshot FromPose(SmplxPose pose, byte lod, uint seq)
    {
        ushort flags = Constants.HasRoot;
        if (lod >= Constants.LodMedium)
            flags |= Constants.HasFace | Constants.HasHands | Constants.HasExpr;
        if (pose.Teleport)
            flags |= Constants.Teleport;
        return new GsAvatarSnapshot
        {
            StreamLod = lod,
            Seq = seq,
            CaptureTick = pose.CaptureTick,
            Flags = flags,
            RootPos = (double[])pose.RootPos.Clone(),
            RootRot = (double[])pose.RootRot.Clone(),
            Body = (double[])pose.Body.Clone(),
            LeftHand = (double[])pose.LeftHand.Clone(),
            RightHand = (double[])pose.RightHand.Clone(),
            Expr = (double[])pose.Expr.Clone(),
        };
    }
}

public static class StreamCodec
{
    public static byte[] Record(SmplxPose pose, byte lod, uint seq) =>
        Encode(GsAvatarSnapshot.FromPose(pose, lod, seq));

    public static SmplxPose Apply(byte[] data) => Decode(data).ToPose();

    public static byte[] Encode(GsAvatarSnapshot snap)
    {
        using var ms = new MemoryStream();
        using var w = new BinaryWriter(ms);
        w.Write(Constants.ProtocolU8);
        w.Write(snap.StreamLod);
        w.Write(snap.Seq);
        w.Write(snap.CaptureTick);
        w.Write(snap.Flags);
        if ((snap.Flags & Constants.HasRoot) != 0)
        {
            WriteF32(w, snap.RootPos, 3);
            WriteF32(w, snap.RootRot, 4);
        }

        var body = Pad(snap.Body, Constants.BodyDim);
        var lh = Pad(snap.LeftHand, Constants.HandDim);
        var rh = Pad(snap.RightHand, Constants.HandDim);
        var expr = snap.Expr ?? Array.Empty<double>();

        if (snap.StreamLod == Constants.LodFull)
        {
            WriteF16(w, body);
            if ((snap.Flags & Constants.HasHands) != 0)
            {
                WriteF16(w, lh);
                WriteF16(w, rh);
            }
            if ((snap.Flags & Constants.HasExpr) != 0)
                WriteF16(w, Pad(expr, Constants.ExprFull));
            w.Flush();
            return ms.ToArray();
        }

        if (snap.StreamLod == Constants.LodLow)
        {
            var low = new double[Constants.LowBodyIndices.Length];
            for (int i = 0; i < low.Length; i++)
                low[i] = body[Constants.LowBodyIndices[i]];
            WriteI16(w, QuantizeAa(low));
            w.Flush();
            return ms.ToArray();
        }

        WriteI16(w, QuantizeAa(body));
        if ((snap.Flags & Constants.HasHands) != 0)
        {
            WriteI16(w, QuantizeAa(lh));
            WriteI16(w, QuantizeAa(rh));
        }
        if ((snap.Flags & Constants.HasExpr) != 0)
        {
            int n = snap.StreamLod == Constants.LodMedium ? Constants.ExprMedium : Constants.ExprFull;
            WriteI16(w, QuantizeExpr(Pad(expr, n)));
        }
        w.Flush();
        return ms.ToArray();
    }

    public static GsAvatarSnapshot Decode(byte[] data)
    {
        using var ms = new MemoryStream(data);
        using var r = new BinaryReader(ms);
        var snap = new GsAvatarSnapshot
        {
            Protocol = r.ReadByte(),
            StreamLod = r.ReadByte(),
            Seq = r.ReadUInt32(),
            CaptureTick = r.ReadUInt32(),
            Flags = r.ReadUInt16(),
        };
        if (snap.Protocol != Constants.ProtocolU8)
            throw new InvalidDataException($"unsupported protocol {snap.Protocol}");
        if ((snap.Flags & Constants.HasRoot) != 0)
        {
            snap.RootPos = ReadF32(r, 3);
            snap.RootRot = ReadF32(r, 4);
        }
        if (snap.StreamLod == Constants.LodFull)
        {
            snap.Body = ReadF16(r, Constants.BodyDim);
            if ((snap.Flags & Constants.HasHands) != 0)
            {
                snap.LeftHand = ReadF16(r, Constants.HandDim);
                snap.RightHand = ReadF16(r, Constants.HandDim);
            }
            if ((snap.Flags & Constants.HasExpr) != 0)
                snap.Expr = ReadF16(r, Constants.ExprFull);
            return snap;
        }
        if (snap.StreamLod == Constants.LodLow)
        {
            var q = ReadI16(r, Constants.LowBodyIndices.Length);
            var body = new double[Constants.BodyDim];
            var aa = DequantizeAa(q);
            for (int i = 0; i < Constants.LowBodyIndices.Length; i++)
                body[Constants.LowBodyIndices[i]] = aa[i];
            snap.Body = body;
            return snap;
        }
        snap.Body = DequantizeAa(ReadI16(r, Constants.BodyDim));
        if ((snap.Flags & Constants.HasHands) != 0)
        {
            snap.LeftHand = DequantizeAa(ReadI16(r, Constants.HandDim));
            snap.RightHand = DequantizeAa(ReadI16(r, Constants.HandDim));
        }
        if ((snap.Flags & Constants.HasExpr) != 0)
        {
            int n = snap.StreamLod == Constants.LodMedium ? Constants.ExprMedium : Constants.ExprFull;
            var e = DequantizeExpr(ReadI16(r, n));
            var expr = new double[Constants.ExprFull];
            Array.Copy(e, expr, n);
            snap.Expr = expr;
        }
        return snap;
    }

    static double[] Pad(double[] src, int n)
    {
        var d = new double[n];
        if (src != null)
            Array.Copy(src, d, Math.Min(src.Length, n));
        return d;
    }

    static short[] QuantizeAa(double[] rad)
    {
        var q = new short[rad.Length];
        for (int i = 0; i < rad.Length; i++)
            q[i] = ClipI16(rad[i] * Constants.AaScale);
        return q;
    }

    static double[] DequantizeAa(short[] q)
    {
        var d = new double[q.Length];
        for (int i = 0; i < q.Length; i++)
            d[i] = q[i] / Constants.AaScale;
        return d;
    }

    static short[] QuantizeExpr(double[] v)
    {
        var q = new short[v.Length];
        for (int i = 0; i < v.Length; i++)
            q[i] = ClipI16(v[i] * Constants.ExprScale);
        return q;
    }

    static double[] DequantizeExpr(short[] q)
    {
        var d = new double[q.Length];
        for (int i = 0; i < q.Length; i++)
            d[i] = q[i] / Constants.ExprScale;
        return d;
    }

    static short ClipI16(double x)
    {
        var r = Math.Round(x);
        if (r > 32767) return 32767;
        if (r < -32767) return -32767;
        return (short)r;
    }

    static void WriteF32(BinaryWriter w, double[] v, int n)
    {
        for (int i = 0; i < n; i++)
            w.Write((float)v[i]);
    }

    static double[] ReadF32(BinaryReader r, int n)
    {
        var v = new double[n];
        for (int i = 0; i < n; i++)
            v[i] = r.ReadSingle();
        return v;
    }

    static void WriteI16(BinaryWriter w, short[] v)
    {
        foreach (var s in v)
            w.Write(s);
    }

    static short[] ReadI16(BinaryReader r, int n)
    {
        var v = new short[n];
        for (int i = 0; i < n; i++)
            v[i] = r.ReadInt16();
        return v;
    }

    static void WriteF16(BinaryWriter w, double[] v)
    {
        foreach (var x in v)
            w.Write(BitConverter.HalfToUInt16Bits((Half)(float)x));
    }

    static double[] ReadF16(BinaryReader r, int n)
    {
        var v = new double[n];
        for (int i = 0; i < n; i++)
            v[i] = (float)BitConverter.UInt16BitsToHalf(r.ReadUInt16());
        return v;
    }
}

File: experimental/python/gsavatar/stream.py
"""GsAvatarStream — Building-Blocks analogue (Topic 4).

RecordGsStreamData(lod) / ApplyGsStreamData(bytes). Transport-agnostic payload.
"""

from __future__ import annotations

import struct
import uuid
from dataclasses import dataclass, field
from typing import Optional

import numpy as np

from .constants import (
    AA_SCALE,
    BODY_DIM,
    COORD_Y_UP_LEFT,
    EXPR_FULL,
    EXPR_MEDIUM,
    EXPR_SCALE,
    HAND_DIM,
    HAS_EXPR,
    HAS_FACE,
    HAS_HANDS,
    HAS_ROOT,
    LOD_FULL,
    LOD_HIGH,
    LOD_LOW,
    LOD_MEDIUM,
    LOW_BODY_INDICES,
    PROTOCOL_U8,
    PROTOCOL_VERSION,
    RIG_SMPLX,
    TELEPORT,
)
from .pose import SmplxPose


def _clip_i16(x: np.ndarray) -> np.ndarray:
    return np.clip(np.rint(x), -32767, 32767).astype(np.int16)


def quantize_aa(rad: np.ndarray) -> np.ndarray:
    return _clip_i16(np.asarray(rad, dtype=np.float64) * AA_SCALE)


def dequantize_aa(q: np.ndarray) -> np.ndarray:
    return np.asarray(q, dtype=np.float64) / AA_SCALE


def quantize_expr(v: np.ndarray) -> np.ndarray:
    return _clip_i16(np.asarray(v, dtype=np.float64) * EXPR_SCALE)


def dequantize_expr(q: np.ndarray) -> np.ndarray:
    return np.asarray(q, dtype=np.float64) / EXPR_SCALE


def _pack_i16(arr: np.ndarray) -> bytes:
    return np.asarray(arr, dtype="<i2").tobytes()


def _pack_f16(arr: np.ndarray) -> bytes:
    return np.asarray(arr, dtype="<f2").tobytes()


def _unpack_i16(data: bytes, n: int, offset: int) -> tuple[np.ndarray, int]:
    raw = data[offset : offset + n * 2]
    if len(raw) != n * 2:
        raise ValueError(f"truncated int16[{n}] at {offset}")
    return np.frombuffer(raw, dtype="<i2").copy(), offset + n * 2


def _unpack_f16(data: bytes, n: int, offset: int) -> tuple[np.ndarray, int]:
    raw = data[offset : offset + n * 2]
    if len(raw) != n * 2:
        raise ValueError(f"truncated f16[{n}] at {offset}")
    return np.frombuffer(raw, dtype="<f2").astype(np.float64), offset + n * 2


@dataclass
class GsAvatarSpawn:
    protocol: int = PROTOCOL_VERSION
    asset_id: bytes = field(default_factory=lambda: uuid.uuid4().bytes)
    rig: int = RIG_SMPLX
    lod_caps: int = 0x0F  # all LODs
    coord_frame: int = COORD_Y_UP_LEFT
    n_gauss: int = 0
    n_bones: int = 0
    n_betas: int = 16
    rest_pose_url: str = ""

    def encode(self) -> bytes:
        url = self.rest_pose_url.encode("utf-8")
        aid = self.asset_id if len(self.asset_id) == 16 else uuid.UUID(bytes=self.asset_id).bytes
        return (
            struct.pack(
                "<H16sBBBIHHH",
                self.protocol,
                aid,
                self.rig,
                self.lod_caps,
                self.coord_frame,
                self.n_gauss,
                self.n_bones,
                self.n_betas,
                len(url),
            )
            + url
        )

    @classmethod
    def decode(cls, data: bytes) -> "GsAvatarSpawn":
        header_size = struct.calcsize("<H16sBBBIHHH")
        if len(data) < header_size:
            raise ValueError("spawn too short")
        proto, aid, rig, lod_caps, coord, n_gauss, n_bones, n_betas, url_len = struct.unpack(
            "<H16sBBBIHHH", data[:header_size]
        )
        url = data[header_size : header_size + url_len].decode("utf-8")
        return cls(
            protocol=proto,
            asset_id=aid,
            rig=rig,
            lod_caps=lod_caps,
            coord_frame=coord,
            n_gauss=n_gauss,
            n_bones=n_bones,
            n_betas=n_betas,
            rest_pose_url=url,
        )


@dataclass
class GsAvatarSnapshot:
    protocol: int = PROTOCOL_U8
    stream_lod: int = LOD_MEDIUM
    seq: int = 0
    capture_tick: int = 0
    flags: int = HAS_ROOT | HAS_FACE | HAS_HANDS | HAS_EXPR
    root_pos: np.ndarray = field(default_factory=lambda: np.zeros(3))
    root_rot: np.ndarray = field(default_factory=lambda: np.array([0.0, 0.0, 0.0, 1.0]))
    body: np.ndarray = field(default_factory=lambda: np.zeros(BODY_DIM))
    left_hand: np.ndarray = field(default_factory=lambda: np.zeros(HAND_DIM))
    right_hand: np.ndarray = field(default_factory=lambda: np.zeros(HAND_DIM))
    expr: np.ndarray = field(default_factory=lambda: np.zeros(EXPR_FULL))

    def to_pose(self) -> SmplxPose:
        expr = self.expr
        if self.stream_lod == LOD_MEDIUM:
            padded = np.zeros(EXPR_FULL)
            n = min(EXPR_MEDIUM, expr.size)
            padded[:n] = expr[:n]
            expr = padded
        elif self.stream_lod == LOD_LOW:
            expr = np.zeros(EXPR_FULL)
        return SmplxPose(
            body=self.body,
            left_hand=self.left_hand,
            right_hand=self.right_hand,
            expr=expr,
            root_pos=self.root_pos,
            root_rot=self.root_rot,
            capture_tick=self.capture_tick,
            teleport=bool(self.flags & TELEPORT),
        )

    @classmethod
    def from_pose(cls, pose: SmplxPose, lod: int, seq: int) -> "GsAvatarSnapshot":
        flags = HAS_ROOT
        if lod >= LOD_MEDIUM:
            flags |= HAS_FACE | HAS_HANDS | HAS_EXPR
        elif lod == LOD_HIGH or lod == LOD_FULL:
            flags |= HAS_FACE | HAS_HANDS | HAS_EXPR
        if pose.teleport:
            flags |= TELEPORT
        return cls(
            stream_lod=lod,
            seq=seq,
            capture_tick=int(pose.capture_tick),
            flags=flags,
            root_pos=pose.root_pos.copy(),
            root_rot=pose.root_rot.copy(),
            body=pose.body.copy(),
            left_hand=pose.left_hand.copy(),
            right_hand=pose.right_hand.copy(),
            expr=pose.expr.copy(),
        )


class StreamCodec:
    """RecordGsStreamData / ApplyGsStreamData."""

    @staticmethod
    def record(pose: SmplxPose, lod: int, seq: int) -> bytes:
        return StreamCodec.encode(GsAvatarSnapshot.from_pose(pose, lod, seq))

    @staticmethod
    def apply(data: bytes) -> SmplxPose:
        return StreamCodec.decode(data).to_pose()

    @staticmethod
    def encode(snap: GsAvatarSnapshot) -> bytes:
        lod = int(snap.stream_lod)
        if lod not in (LOD_LOW, LOD_MEDIUM, LOD_HIGH, LOD_FULL):
            raise ValueError(f"bad lod {lod}")
        flags = int(snap.flags)
        buf = bytearray(
            struct.pack(
                "<BBIIH",
                PROTOCOL_U8,
                lod,
                int(snap.seq) & 0xFFFFFFFF,
                int(snap.capture_tick) & 0xFFFFFFFF,
                flags & 0xFFFF,
            )
        )
        if flags & HAS_ROOT:
            pos = np.asarray(snap.root_pos, dtype=np.float64).reshape(3)
            rot = np.asarray(snap.root_rot, dtype=np.float64).reshape(4)
            buf += struct.pack("<7f", *pos.tolist(), *rot.tolist())

        body = np.zeros(BODY_DIM, dtype=np.float64)
        b_in = np.asarray(snap.body, dtype=np.float64).reshape(-1)
        body[: min(BODY_DIM, b_in.size)] = b_in[:BODY_DIM]

        lh = np.zeros(HAND_DIM)
        rh = np.zeros(HAND_DIM)
        li = np.asarray(snap.left_hand, dtype=np.float64).reshape(-1)
        ri = np.asarray(snap.right_hand, dtype=np.float64).reshape(-1)
        lh[: min(HAND_DIM, li.size)] = li[:HAND_DIM]
        rh[: min(HAND_DIM, ri.size)] = ri[:HAND_DIM]

        expr = np.asarray(snap.expr, dtype=np.float64).reshape(-1)

        if lod == LOD_FULL:
            buf += _pack_f16(body)
            if flags & HAS_HANDS:
                buf += _pack_f16(lh)
                buf += _pack_f16(rh)
            if flags & HAS_EXPR:
                e = np.zeros(EXPR_FULL)
                e[: min(EXPR_FULL, expr.size)] = expr[:EXPR_FULL]
                buf += _pack_f16(e)
            return bytes(buf)

        if lod == LOD_LOW:
            buf += _pack_i16(quantize_aa(body[list(LOW_BODY_INDICES)]))
            return bytes(buf)

        buf += _pack_i16(quantize_aa(body))
        if flags & HAS_HANDS:
            buf += _pack_i16(quantize_aa(lh))
            buf += _pack_i16(quantize_aa(rh))
        if flags & HAS_EXPR:
            if lod == LOD_MEDIUM:
                e = np.zeros(EXPR_MEDIUM)
                e[: min(EXPR_MEDIUM, expr.size)] = expr[:EXPR_MEDIUM]
                buf += _pack_i16(quantize_expr(e))
            else:
                e = np.zeros(EXPR_FULL)
                e[: min(EXPR_FULL, expr.size)] = expr[:EXPR_FULL]
                buf += _pack_i16(quantize_expr(e))
        return bytes(buf)

    @staticmethod
    def decode(data: bytes) -> GsAvatarSnapshot:
        if len(data) < 12:
            raise ValueError("snapshot shorter than header")
        proto, lod, seq, tick, flags = struct.unpack_from("<BBIIH", data, 0)
        if proto != PROTOCOL_U8:
            raise ValueError(f"unsupported protocol {proto}")
        off = 12
        root_pos = np.zeros(3)
        root_rot = np.array([0.0, 0.0, 0.0, 1.0])
        if flags & HAS_ROOT:
            if len(data) < off + 28:
                raise ValueError("truncated root")
            vals = struct.unpack_from("<7f", data, off)
            root_pos = np.array(vals[:3], dtype=np.float64)
            root_rot = np.array(vals[3:], dtype=np.float64)
            off += 28

        body = np.zeros(BODY_DIM)
        lh = np.zeros(HAND_DIM)
        rh = np.zeros(HAND_DIM)
        expr = np.zeros(EXPR_FULL)

        if lod == LOD_FULL:
            body, off = _unpack_f16(data, BODY_DIM, off)
            if flags & HAS_HANDS:
                lh, off = _unpack_f16(data, HAND_DIM, off)
                rh, off = _unpack_f16(data, HAND_DIM, off)
            if flags & HAS_EXPR:
                expr, off = _unpack_f16(data, EXPR_FULL, off)
        elif lod == LOD_LOW:
            q, off = _unpack_i16(data, len(LOW_BODY_INDICES), off)
            body[list(LOW_BODY_INDICES)] = dequantize_aa(q)
        else:
            q, off = _unpack_i16(data, BODY_DIM, off)
            body = dequantize_aa(q)
            if flags & HAS_HANDS:
                q, off = _unpack_i16(data, HAND_DIM, off)
                lh = dequantize_aa(q)
                q, off = _unpack_i16(data, HAND_DIM, off)
                rh = dequantize_aa(q)
            if flags & HAS_EXPR:
                n = EXPR_MEDIUM if lod == LOD_MEDIUM else EXPR_FULL
                q, off = _unpack_i16(data, n, off)
                expr[:n] = dequantize_expr(q)

        return GsAvatarSnapshot(
            protocol=proto,
            stream_lod=lod,
            seq=seq,
            capture_tick=tick,
            flags=flags,
            root_pos=root_pos,
            root_rot=root_rot,
            body=body,
            left_hand=lh,
            right_hand=rh,
            expr=expr,
        )


def snapshot_nbytes(lod: int, flags: Optional[int] = None) -> int:
    """Exact encoded size for zeroed pose (used by bandwidth tests)."""
    pose = SmplxPose.identity()
    if flags is None:
        return len(StreamCodec.record(pose, lod, 0))
    snap = GsAvatarSnapshot.from_pose(pose, lod, 0)
    snap.flags = flags
    return len(StreamCodec.encode(snap))
